Charlesmchen/split gifs

charlesmchen opened the pull request:
PTAL @michaelkirk

michaelkirk commented on Signal/src/network/GiphyDownloader.swift:
```diff
@@ -227,93 +434,220 @@ extension URLSessionTask {
+                })
+
+                task.resume()
+                return
```
Do we need to kick the run loop here?

Else won't we potentially leave the queue hanging or less-than-full?

And same question for the end of this method?

charlesmchen replied:
Already fixed!

charlesmchen commented on Signal/src/ViewControllers/GifPicker/GifPickerViewController.swift:
```diff
@@ -147,6 +147,7 @@ class GifPickerViewController: OWSViewController, UISearchBarDelegate, UICollect
     private func createViews() {
 
         view.backgroundColor = UIColor.white
+        self.automaticallyAdjustsScrollViewInsets = false
```
`UIViewController`'s default behavior is to sometimes manipulate the insets of any scroll view that is the root view or a direct descendent thereof - the intent is to get scrolling content to layout out properly vis-a-vis navigation headers etc.  The problem is that UIKit doesn't always guess right about when you want this behavior or not.  In our case, it was breaking scroll behavior in this view.  Trying scrolling up past the top of the GIF content.

michaelkirk replied:
:+1: I've actually seen that once before and forgot. Thanks for addressing it. Can you include a comment?

charlesmchen replied:
Done.

charlesmchen commented on Signal/src/network/GiphyDownloader.swift:
```diff
@@ -10,11 +10,79 @@ enum GiphyRequestPriority {
+        var bytesToIgnore = Int(redundantLength)
+        for data in datas {
+            if data.count <= bytesToIgnore {
+                bytesToIgnore -= data.count
```
I was being clever and trying to avoid doing redundant work:

* merging segment data
* merging asset data.

To avoid the first merge, I was storing a list of segment data pieces.  The code's a lot simpler if we just do two merges though so I've done that.

charlesmchen commented on Signal/src/network/GiphyDownloader.swift:
```diff
@@ -28,15 +96,137 @@ enum GiphyRequestPriority {
+            return 0
+        }
+
+        let k1MB: UInt = 1024 * 1024
```
Great question - I'll ask Moxie.

michaelkirk commented on Signal/src/network/GiphyDownloader.swift:
```diff
@@ -136,41 +163,46 @@ enum GiphyAssetRequestState: UInt {
-                redundantLength = segmentStart + segmentLength - fileSize
-                segmentStart = fileSize - segmentLength
+            if segmentStart + segmentLength > contentLength {
+                redundantLength = segmentStart + segmentLength - contentLength
```
nit: The logic currently works out, but it seems counterintuitive that `redundantLength` will be negative for all but the last segments. I'd expect it to be 0 in those cases.

charlesmchen replied:
Already reworked.  See the new arrangement.

michaelkirk commented on Signal/src/network/GiphyDownloader.swift:
```diff
@@ -522,7 +585,7 @@ extension URLSessionTask {
             guard !assetRequest.wasCancelled else {
                 // Discard the cancelled asset request and try again.
-                self.processRequestQueue()
+                self.removeAssetRequestFromQueue(assetRequest: assetRequest)
```
Don't we still want to `processRequestQueue()` so we can try the next one?

charlesmchen replied:
`removeAssetRequestFromQueue` calls `processRequestQueue`

michaelkirk commented on Signal/src/network/GiphyDownloader.swift:
```diff
@@ -557,34 +620,46 @@ extension URLSessionTask {
 
-                let task = downloadSession.dataTask(with:request, completionHandler: { [weak self] _, response, error -> Void in
+                let task = downloadSession.dataTask(with:request, completionHandler: { [weak self] data, response, error -> Void in
+                    if let data = data {
```
style nit: combine these clauses `if let data = data, data.count > 0 {`

charlesmchen replied:
Nice.

michaelkirk commented on Signal/src/network/GiphyDownloader.swift:
```diff
@@ -617,19 +692,31 @@ extension URLSessionTask {
                 case .requestingSize:
                     activeAssetRequestsCount += 1
+                    // Ensure that only N requests are active at a time.
+                    guard activeAssetRequestsCount < kMaxAssetRequestCount else {
```
Naming suggests this should be `<=`

michaelkirk replied:
It seems like this guard belongs in the `case .waiting` instead of here.

I think we want to do what we can to ensure something of a FIFO priority with our GiphyAssets. And requesting content size is a way to insert network requests for Asset which we haven't "started" yet.

charlesmchen replied:
> Naming suggests this should be <=

We're try to find an asset for which we can add a request.  If we're at MAX, we don't want to add another request and can exit early.

charlesmchen replied:
> It seems like this guard belongs in the case .waiting instead of here.

We can abandon our search for a request to start as soon as we hit the limit.  We want this method to be cheap since we call it often.

> I think we want to do what we can to ensure something of a FIFO priority with our GiphyAssets. 

We do honor FIFO, within the limitation that we have an upper bound on the number of requests per asset.  Are you saying we shouldn't do that?  

> And requesting content size is a way to insert network requests for Asset which we haven't "started" yet.

Right under the current arrangement we can have requests for multiple assets at a time.  For example, when the picker opens, it'll immediately do HEAD on the first six assets.  When the first HEAD completes, we'll start a segment download.  etc.

If you think we need a different arrangement let's video chat.

michaelkirk commented on Signal/src/network/GiphyDownloader.swift:
```diff
@@ -639,13 +726,18 @@ extension URLSessionTask {
+                let downloadingSegmentsCount = assetRequest.downloadingSegmentsCount()
+                activeAssetRequestsCount += downloadingSegmentsCount
+                // Ensure that only N segment requests are active per asset at a time.
+                guard downloadingSegmentsCount < kMaxAssetRequestsPerAssetCount else {
```
Naming suggests this should be `<=`

charlesmchen replied:
See above, we want to enforce the limit and not exceed it.

michaelkirk commented on Signal/src/network/GiphyDownloader.swift:
```diff
@@ -639,13 +726,18 @@ extension URLSessionTask {
+                    continue
+                }
+                // Ensure that only N requests are active at a time.
+                guard activeAssetRequestsCount < kMaxAssetRequestCount else {
```
Naming suggests this should be `<=`

charlesmchen replied:
Ditto.

michaelkirk commented on Signal/src/network/GiphyDownloader.swift:
```diff
@@ -46,15 +51,27 @@ class GiphyAssetSegment {
     }
 
     public func mergeData(assetData: NSMutableData) {
+        guard state == .complete else {
```
Not functionally wrong, but why are we using `NSMutableData`?  I'd expect something like:

```
mergeData(assetData: inout Data) {
...
assetData.append(data.subdata(in: start..<end))
...
}

var assetData = Data()
mergeData(assetData: &assetData)
```

charlesmchen replied:
Done.

michaelkirk commented on Signal/src/network/GiphyDownloader.swift:
```diff
@@ -46,15 +51,27 @@ class GiphyAssetSegment {
     }
 
     public func append(data: Data) {
+        guard state == .downloading else {
```
Why does an AssetSegment have multiple datas? 

I expected just one, represented by it's range (whereas the GiphyAsset has multiple data's via it's multiple AssetSegments.)

charlesmchen replied:
Sorry I got interrupted by an interview.  I've just pushed a commit that unifies this.

charlesmchen commented on Signal/src/network/GiphyDownloader.swift:
```diff
@@ -227,93 +434,220 @@ extension URLSessionTask {
+    private func processRequestQueue() {
         AssertIsOnMainThread()
 
         DispatchQueue.main.async {
```
I've reworked this a bit since we only need to do this dispatch async `requestAsset()` and added a comment.

charlesmchen commented:
PTAL @michaelkirk I've just pushed yet another commit.

> In my testing (I'm on a crappy network) I'm seemingly able to get the GifPicker into a state where it stops making progress. 

Fixed, we weren't handling all cancellations correctly.

> See my concerns about missing some recursion calls in processRequestQueue

Fixed.

> the request queue can get pretty backlogged (e.g. I'm seeing hundreds of requests when I quickly scroll to the bottom).

Queue entries (the asset requests) should be cheap and stale ones should be quickly cleared.  

> we don't actually cancel the dataTask until the next time the NSURLSession delegate checks in, which could be a while on a slow network.

I've modified the cancellation logic to cancel the tasks sooner, but I'm not sure if that'll have any impact.

charlesmchen commented on Signal/src/network/GiphyDownloader.swift:
```diff
@@ -77,11 +81,11 @@ class GiphyAssetSegment {
-            let range = NSMakeRange(bytesToIgnore, segmentData.length - bytesToIgnore)
-            let subdata = (segmentData as NSData).subdata(with: range)
+            let range = NSMakeRange(bytesToIgnore, segmentData.count - bytesToIgnore)
+            let subdata = segmentData.subdata(in: range.location..<range.location + range.length)
```
In the next PR!
